packages/liveblocks-core/src/crdts/__tests__/cloning.test.ts

```diff
@@ -0,0 +1,106 @@
+import * as fc from "fast-check";
+
+import {
+  listUpdate,
+  listUpdateInsert,
+  objectUpdate,
+} from "../../__tests__/_updatesUtils";
+import {
+  createSerializedList,
+  createSerializedObject,
+  prepareStorageUpdateTest,
+} from "../../__tests__/_utils";
+import type { LiveList } from "../LiveList";
+import { liveStructure } from "./_arbitraries";
+
+describe("cloning LiveStructures", () => {
+  it("basic cloning logic", async () => {
+    const { root, expectUpdates, room } = await prepareStorageUpdateTest<{
+      list1: LiveList<string>;
+      list2: LiveList<string>;
+    }>([
+      createSerializedObject("0:0", {}),
+      createSerializedList("0:1", "0:0", "list1"),
+    ]);
+
+    const list1 = root.get("list1");
+    list1.push("a");
+    list1.push("b");
+    list1.push("c");
+    root.set("list2", list1.clone());
+
+    room.history.undo();
+    room.history.redo();
+
+    expectUpdates([
+      // List creation
+      [listUpdate(["a"], [listUpdateInsert(0, "a")])],
+      [listUpdate(["a", "b"], [listUpdateInsert(1, "b")])],
+      [listUpdate(["a", "b", "c"], [listUpdateInsert(2, "c")])],
+
+      // Clone
+      [
+        objectUpdate(
+          { list1: ["a", "b", "c"], list2: ["a", "b", "c"] },
+          { list2: { type: "update" } }
+        ),
+      ],
+
+      // Undo
+      [
+        objectUpdate({ list1: ["a", "b", "c"] }, {
+          list2: { type: "delete" },
+        } as any),
+      ],
+
+      // Redo
+      [
+        objectUpdate(
+          { list1: ["a", "b", "c"], list2: ["a", "b", "c"] },
+          { list2: { type: "update" } }
+        ),
+      ],
+    ]);
+  });
+
+  it("large/deep cloning", () =>
+    fc.assert(
+      fc.asyncProperty(
+        liveStructure,
+
+        async (data) => {
+          const { root } = await prepareStorageUpdateTest([
+            createSerializedObject("0:0", {}),
+          ]);
+
+          // Clone "a" to "b"
+          root.set("a", data);
+          root.set("b", data.clone());
+
+          const imm = root.toImmutable();
+          expect(imm.a).toEqual(imm.b);
+        }
+      )
+    ));
```

Comment:
Here's the simplest test explained:

First, we use the `liveStructure` Arbitrary (see above) to generate randomized LiveMap/LiveList/LiveObject structures. (Some deep, some shallow, some simple, some complex, etc.)

Then, we inject it into the root, under key `"a"`. Next, we clone the value and inject it under key "b". Note that inserting the same value into the tree in different places is illegal and would lead to an error, e.g.

```tsx
root.set("a", data);
root.set("b", data);  // ❌ Would error!
```

But not when you clone it—then it's fine:

```tsx
root.set("a", data);
root.set("b", data.clone());  // ✅ Fine!
```

Finally, we check that the cloned tree will produce the same immutable value for keys "a" and "b".